Lib/test/test_asyncio/test_proactor_events.py

```diff
@@ -489,6 +407,191 @@ def test_dont_pause_writing(self):
         self.assertFalse(self.protocol.pause_writing.called)
 
 
+class ProactorSocketTransportWithProtocolTests(test_utils.TestCase, ProactorSocketTransportTestsBase):
+
+    def make_test_protocol(self):
+        self.protocol = test_utils.make_test_protocol(asyncio.Protocol)
+
+    def setUp(self):
+        super().setUp()
+        self.setup_globals()
+
+    def test_ctor(self):
+        fut = self.loop.create_future()
+        tr = self.socket_transport(waiter=fut)
+        test_utils.run_briefly(self.loop)
+        self.assertIsNone(fut.result())
+        self.protocol.connection_made(tr)
+        self.proactor.recv.assert_called_with(self.sock, self.buffer_size)
+
+    def test_loop_reading(self):
+        tr = self.socket_transport()
+        tr._loop_reading()
+        self.loop._proactor.recv.assert_called_with(self.sock, self.buffer_size)
+        self.assertFalse(self.protocol.data_received.called)
+        self.assertFalse(self.protocol.eof_received.called)
+
+    def test_loop_reading_data(self):
+        buf = b'data'
+        res = self.loop.create_future()
+        res.set_result(buf)
+
+        tr = self.socket_transport()
+        tr._read_fut = res
+        tr._loop_reading(res)
+        self.loop._proactor.recv.assert_called_with(self.sock, self.buffer_size)
+        self.protocol.data_received.assert_called_with(buf)
+
+    def test_pause_resume_reading(self):
+        tr = self.socket_transport()
+        futures = []
+        msgs = [b'data1', b'data2', b'data3', b'data4', b'data5', b'']
+        for msg in msgs:
+            f = self.loop.create_future()
+            f.set_result(msg)
+            futures.append(f)
+
+        self.loop._proactor.recv.side_effect = futures
+        self.loop._run_once()
+        self.assertFalse(tr._paused)
+        self.assertTrue(tr.is_reading())
+
+        for msg in msgs[:2]:
+            self.loop._run_once()
+            self.protocol.data_received.assert_called_with(msg)
+
+        tr.pause_reading()
+        tr.pause_reading()
+        self.assertTrue(tr._paused)
+        self.assertFalse(tr.is_reading())
+        for _ in range(10):
+            self.loop._run_once()
+        self.protocol.data_received.assert_called_with(msgs[1])
+
+        tr.resume_reading()
+        tr.resume_reading()
```

Comment:
why this double `pause_reading()` above and then `resume_reading()` here ?

and same question for below in `test_pause_resume_reading`.

Reply:
I did not write this test, I copied it to another place, that's why there is diff here.

Reply:
I am not sure, but that's how it was when I copied the test, only to change some minor stuff inside it after my changes.